### PR TITLE
BERT StackOverflow Fine-Tuning

### PR DESCRIPTION
**Pull request by Readrid**

### CODE REVIEW

File: text_models/datasets/triplet.py [hunk 1]
```diff
@@ -29,24 +29,30 @@ def __init__(
 
         self.triplets = []
         for dupl_series in duplicate_clusters.values():
-            if len(dupl_series) <= 1:
+            n_dupl = len(dupl_series)
+            if n_dupl <= 1:
                 continue
-            for i in range(len(dupl_series)):
+
+            old_negs: Dict[int, int] = dict()
```
Comment (egor-bogomolov) [inline text_models/datasets/triplet.py]: What does this dict store? It is not obvious from its name, and I cannot understand what the check below prevents :crying_cat_face:

File: data_processing/util.py [hunk 2]
```diff
@@ -23,10 +23,14 @@
 NumpyNaN = float
 
 
-def get_corpus(data: pd.DataFrame, sentences: bool = False) -> Union[Section, Corpus]:
+def get_corpus(data: pd.DataFrame, sentences: bool = False) -> Section:
+    def parse_list(str_list):
+        return [] if isinstance(str_list, float) else ast.literal_eval(str_list)
```
Comment (egor-bogomolov) [inline data_processing/util.py:28]: What is the data format that you parse here?

File: data_processing/util.py [hunk 3]
```diff
@@ -23,10 +23,14 @@
...
+
     corpus = []
-    for str_list in data["description"].tolist():
-        corpus.append(ast.literal_eval(str_list))
+    for summ, descr in zip(data.summary, data.description):
+        full_descr = parse_list(summ) + parse_list(descr)
+        corpus.append(full_descr)
```
Comment (egor-bogomolov) [inline data_processing/util.py]: The loop can be replaced with list comp

File: data_processing/preprocess_csv.py [hunk 4]
```diff
@@ -20,10 +20,10 @@ def main():
...
     data = pd.read_csv(config.datasets.full)
 
-    data = data.dropna(axis=0, subset=["description"])
-    data = data.reset_index(drop=True)
-    data["description"] = data["description"].apply(preprocess)
-    data = data.dropna(axis=0, subset=["description"])
+    data["summary"] = data["summary"].map(preprocess, na_action="ignore")
+    data["description"] = data["description"].map(preprocess, na_action="ignore")
+
+    data = data[~(data.summary.isnull() & data.description.isnull())]
```
Comment (egor-bogomolov) [inline data_processing/preprocess_csv.py]: It's up to you, but I feel that `df.x.notnull() | df.y.notnull()` is a bit more readable

File: approaches/abstract_approach.py [hunk 5]
```diff
@@ -84,48 +87,74 @@ def save_results(self, save_to_path: str, model_name: str, plot: bool = False):
...
         self.embeddings = model.get_embeddings(self.train_corpus)
         self.test_embs = model.get_embeddings(self.test_corpus)
+        self.train_copy = self.train.copy()
 
         self.setup_approach()
 
         self.test_size = 0
-        self.true_positive = np.zeros(len(topns))
+        self.true_positive_at_k = np.zeros(len(topks))
+        self.ave_precision_at_k = np.zeros(len(topks))
```
Comment (egor-bogomolov) [inline approaches/abstract_approach.py]: Common short for "average" is "avg"

File: approaches/abstract_approach.py [hunk 6]
```diff
@@ -138,3 +167,21 @@ def get_duplicated_ids(self, query_num: int, topn: int) -> np.ndarray:
     @abstractmethod
     def update_history(self, query_num: int):
         raise NotImplementedError()
+
+    def __get_relevant_reports_num(self) -> Dict[str, int]:
+        result: Dict[str, int] = dict()
```
Comment (egor-bogomolov) [inline approaches/abstract_approach.py]: You can use `collections.Counter` and avoid later checking key presence in the dict. If the stored data was something non-integer, you could've used `collections.defaultdict`

File: approaches/abstract_approach.py [hunk 7]
```diff
@@ -138,3 +167,21 @@ def get_duplicated_ids(self, query_num: int, topn: int) -> np.ndarray:
...
+
+        query_ids = self.test.id.tolist()
+        train_disc_ids = self.train.disc_id.tolist()
+        test_disc_ids = self.test.disc_id.tolist()
```
Comment (egor-bogomolov) [inline approaches/abstract_approach.py]: What is the structure of these dataframes? It looks like the iteration might be made more efficiently

**Review by egor-bogomolov — CHANGES_REQUESTED** [review]

File: text_models/datasets/triplet.py (hunk 1 above)
Comment (Readrid) [reply text_models/datasets/triplet.py]: It stores pairs of bug reports that have already been used together. I renamed this to `used_pos_negs_pairs`

File: data_processing/util.py (hunk 2 above)
Comment (Readrid) [reply data_processing/util.py:28]: This is a python list of list of tokens converted to string: `[ ["hello", "world"], ["word1", ...], ...]`

File: data_processing/util.py (hunk 3 above)
Comment (Readrid) [reply data_processing/util.py]: Ok, thank you!

File: data_processing/preprocess_csv.py (hunk 4 above)
Comment (Readrid) [reply data_processing/preprocess_csv.py]: Ok, I think your method is better :)

File: approaches/abstract_approach.py (hunk 5 above)
Comment (Readrid) [reply approaches/abstract_approach.py]: Of course, fixed this

File: approaches/abstract_approach.py (hunk 6 above)
Comment (Readrid) [reply approaches/abstract_approach.py]: Ok, now I use `collections.Counter`

File: approaches/abstract_approach.py (hunk 7 above)
Comment (Readrid) [reply approaches/abstract_approach.py]: The structure is something like this:

|  id  | ... |   summary   |  ...  |  description  |  ...  | disc_id |
|:----:|-----|:-----------:|:-----:|:-------------:|:-----:|:-------:|
| ...  | ... |     ...     |  ...  |      ...      |  ...  |   ...   |
|  33  | ... | bug example |  ...  | this is a bug |  ...  |   11    |

I tried to optimise this method in this way:

```python
  def __get_relevant_reports_num(self) -> Counter:
      result: Counter = Counter()
      train_disc_ids = self.train.disc_id.tolist()
      def collect_relevant_reports(query):
          for train_disc_id in train_disc_ids:
              result[query.id] += (query.disc_id == train_disc_id)
          train_disc_ids.append(query.disc_id)

      self.test.apply(collect_relevant_reports, axis=1)

      return result
```

but it starts to work much slower. So I left the old implementation.

Comment (Readrid) [reply approaches/abstract_approach.py]: Ok, I did some optimizations, now it looks like this:
```python
    def __get_relevant_reports_num(self) -> Counter:
        result: Counter = Counter()

        test_disc_ids = self.test.disc_id.tolist()
        count_train: Counter = Counter(self.train.disc_id)

        for i, query_id in enumerate(self.test.id):
            result[query_id] += count_train[test_disc_ids[i]]
            count_train[test_disc_ids[i]] += 1

        return result
```

File: data_processing/util.py (hunk 2 above)
Comment (Readrid) [reply data_processing/util.py:28]: This is a python list of list of tokens converted to string: `[["hello", "world"], ["word1", ...], ...]`

**Pull request merged**